fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/BaseScalarOperatorShuttle.java

```diff
@@ -0,0 +1,216 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Inc.
+
+package com.starrocks.sql.optimizer.rewrite;
+
+import com.clearspring.analytics.util.Lists;
+import com.starrocks.sql.optimizer.operator.scalar.ArrayElementOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ArrayOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ArraySliceOperator;
+import com.starrocks.sql.optimizer.operator.scalar.BetweenPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.BinaryPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CallOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CaseWhenOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CastOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CompoundPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ConstantOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ExistsPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.InPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.IsNullPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.LikePredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.PredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperatorVisitor;
+
+import java.util.List;
+
+public class BaseScalarOperatorShuttle extends ScalarOperatorVisitor<ScalarOperator, Void> {
```

Comment:
the class just use for clone? why don't use clone directly?

Reply:
this is a base class which used to generate a new scalarOperator when we want to replace a node in the old scalarOperator.

Reply:
so why don't use ScalarOperator.clone() directly? It seems that the works are the same

Reply:
We shoud never directly use this class. If you don't override its visit method, this shuttle just traverse the operator tree and do nothing even copy the operator. But when you want change some nodes in the original operator tree in your way, you can extend this class implement your own shuttle. If you changed some node, the shuttle will detect it and generate a new operator tree to you. I think it helps to avoid duplication work when you want to rewrite the operator tree.

Reply:
If I want to remove, not replace, some of the inner nodes from the originl tree, can I make it though this class?

Reply:
Technically, I think it is feasible but logic maybe complex. If you remove a node, his father's child decrease and this may affect the type of his father.

Reply:
Fine, I got it, it's means all the nodes above it will be re-new when a node is modified, but add comment and give a explicit name plz, help others can use it quickly

Reply:
yeh, I add some comments and explain the shuttle meaning.